packages/gatsby-plugin-sitemap/src/gatsby-node.js

```diff
@@ -1,80 +1,139 @@
+  } catch (err) {
+    reporter.panic(err)
+  }
+}
+
 exports.onPostBuild = async (
-  { graphql, pathPrefix, basePath = pathPrefix },
+  { graphql, reporter, pathPrefix },
   pluginOptions
 ) => {
-  const options = { ...pluginOptions }
-  delete options.plugins
-  delete options.createLinkInHead
+  global.reporter = reporter
 
   const {
-    query,
-    serialize,
     output,
-    exclude,
-    hostname,
+    entryLimit,
+    query,
+    excludes,
     resolveSiteUrl,
-    ...rest
-  } = {
-    ...defaultOptions,
-    ...options,
+    resolvePagePath,
+    resolvePages,
+    filterPages,
+    serialize,
+  } = await validateOptions(pluginOptions).catch(err => {
+    reporter.panic(err)
+  })
+
+  const { data: queryRecords } = await graphql(query)
+
+  reporter.verbose(
+    `${reporterPrefix} Query Results` + JSON.stringify(queryRecords, null, 2)
+  )
+
+  const allPages = resolvePages(queryRecords)
+  const siteUrl = resolveSiteUrl(queryRecords)
+
+  if (!Array.isArray(allPages)) {
+    reporter.panic(
+      `The \`resolvePages\` function did not return an array, if you wrote a custom query you may need to provide a custom \`resolvePages\` function.`
+    )
   }
 
-  const saved = path.join(publicPath, output)
+  reporter.verbose(
+    `${reporterPrefix} Filtering ${allPages.length} pages based on ${excludes.length} excludes`
+  )
+
+  const filteredPages = allPages
+    .filter(page => {
+      const result = !defaultExcludes.some(exclude => {
+        try {
+          return defaultFilterPages(page, exclude, {
+            minimatch,
+            withoutTrailingSlash,
+            resolvePagePath,
+          })
+        } catch (err) {
+          reporter.panic(
+            `${reporterPrefix} default filtering of pages failed.This shouldn't really be possible, please report a bug.`
+          )
+
+          return err
+        }
+      })
+
+      if (!result) {
+        reporter.verbose(
+          `${reporterPrefix} Default filter excluded page ${resolvePagePath(
+            page
+          )}`
+        )
+      }
+
+      return result
+    })
+    .filter(page => {
+      const result = !excludes.some(exclude => {
+        try {
+          return filterPages(page, exclude, {
+            minimatch,
+            withoutTrailingSlash,
+            resolvePagePath,
+          })
+        } catch (err) {
+          reporter.panic(
+            `${reporterPrefix} custom page filtering failed. If you've customized your excludes you may need to provide a custom "filterPages" function in your config.`
+          )
+
+          return err
+        }
+      })
 
-  // Paths we're excluding...
-  const excludeOptions = exclude.concat(defaultOptions.exclude)
+      if (!result) {
+        reporter.verbose(
+          `${reporterPrefix} Custom filtering excluded page ${resolvePagePath(
+            page
+          )}`
+        )
+      }
 
-  const queryRecords = await graphql(query)
+      return result
+    })
```

Comment:
Can we extract this into a sub-function so the flow of how a sitemap is generated is easier to understand?

You probably want to combine 2 filters int 1 as this might become slow when you have 100k pages.

Reply:
Yes I can break this out, good call. The reason I specifically did two filters was so folks could easily write a custom filter without breaking and having to reimplement  the default filter. I'll do some thinking on how to make that more performant.

Reply:
I assumed it was for that reason but we can still do it in one filter loop, no?

Reply:
Probably. 🤪

Reply:
Okay, I've moved this code out, moved it to a single filter and added some logic to reduce filter checks for the default page filters. More could probably be done but this will be much better.